### ISSUE TITLE
Attribute-guided image generation from layout

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#114]
*BMVC 2020*

# TL;DR

- **Layout2Im + Attribute + Global Context Encoder + SPADE + Shifted Output**

# Motivation

- Semantically controllable complex scene generation (using attribute)
- Spatially equivariant model (consistency)

# Previous Work

- Layout2Im: Encoder-Fuser(Conv-LSTM)-Decoder, Latent Regression
- LostGAN: Spatially Adaptive Normalization + Embedded Label Projection (Discriminator)

# Method

![image](https://user-images.githubusercontent.com/46308407/130718467-276a5c48-84b3-4abc-833b-61fa068ac7f7.png)

## Incorporating (optional) Attribute

- Word embedding (label), attribute vector (multi-hot), object latent are concatenated.
- Spatially aligned feature map is generated by simply filling in corresponding bbox with concatenated feature vectors.
- Note that attribute vector can be zero, or sampled from prior distribution (random attribute)
- To *disentangle* attribute from object latent (that models variation in appearance) and label word embedding, randomly choose a subset of training samples and replace GT attribute with new attributes sampled from corresponding object class attribute frequency distribution.

## Global Context Encoder
![image](https://user-images.githubusercontent.com/46308407/130718861-9026a127-e4cb-487b-b235-353fc0889564.png)

## Objective

![image](https://user-images.githubusercontent.com/46308407/130719053-f453ae9f-e741-430d-8d34-1737603d8a65.png)

- Generates three outputs: *recon, random, and shift*.
- Recon uses **posterior** object latent inferred from **Object Estimator**, and L1 reconstruction loss is applied.
- Random uses **prior (standard normal)** latent code.
- Shift uses same **prior** latent code, but bbox is shifted *horizontally*.

![image](https://user-images.githubusercontent.com/46308407/130719067-c57bd0d6-cff8-4d72-be79-211293bfbc7c.png)
![image](https://user-images.githubusercontent.com/46308407/130719082-f7190094-6744-4bcc-bc13-59ca6e0050f5.png)

- Note that attribute classification is added on top of *Layout2Im*
- KL enforces estimated object latent posterior distribution to be close to unit Gaussian.

# Experiments

## Datasets
- Visual Genome
- ***COCO is not used since it does not provide attributes***

## Baseline

![image](https://user-images.githubusercontent.com/46308407/130719259-c94a344c-1c26-4395-9e20-c8e63dd4709f.png)

![image](https://user-images.githubusercontent.com/46308407/130719245-b8fa5d21-0b80-482b-9591-426565612a83.png)

![image](https://user-images.githubusercontent.com/46308407/130719289-fe0dee06-c9e8-4fbc-b96f-07896fa9c2aa.png)

## Ablation
![image](https://user-images.githubusercontent.com/46308407/130719301-3ec8a109-2913-4f13-a9bf-724b4c892a5d.png)

## Analysis
![image](https://user-images.githubusercontent.com/46308407/130719319-fbf6ba5c-3256-43a0-bf13-1140301399e7.png)

# Comments
- Idea to incorporate attribute to L2I task is very welcomed by me.
- Wonder why they chose Layout2Im as their baseline. LostGAN is much simpler in its design, and performs better.
- Experiment is relatively weak as the only dataset they used was Visual Genome. Wonder if they could have leveraged other datasets. (CelebA has attributes, but images are too iconic and lacks diversity. Maybe outpaint on celebA to perturb face location and employ its attributes?)
